### PR TITLE
onboard: request invites + existing race fix

### LINKED ISSUE #695: Stuck on "Creating account..."
## Summary

**UI sometimes gets stuck during account creation.** Account creation succeeds.

## Investigation

- Overly spammy log. Note repeated `ONBOARDING] chainId baseSepolia`
- Onboarding logic too entwined with react hooks.
- Start: `[ACTION] 0ms: idle > loading Creating account...`
- No further `[ACTION] ...` logs even though creation succeeds.

**This bug should be fixed by decoupling onboarding UI from account creation logic.**

```
[ONBOARDING] chainId baseSepolia
 LOG  [ACTION] 0ms: idle > loading Creating account...
 LOG  [ONBOARDING] create account test23 idle
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [TRPC] fetching http://192.168.1.22:3000/chain/84532/deployWallet?batch=1, timeout 60000ms {"body": "{\"0\":{\"name\":\"test23\",\"pubKeyHex\":\"0x3059301306072a8648ce3d020106082a8648ce3d0301070342000475092d95ffec5cea8e599d852262c62a45e71b533bbf89c6b55297b2829e5a6257e5aa86e879da417e5a28abb41c3587062b8dc30f8048097f2732de01551847\",\"inviteLink\":\"http://localhost:3001/link/invite/testnet\"}}", "headers": {"content-type": "application/json", "x-daimo-platform": "ios 17.0.1", "x-daimo-version": "1.5.0 #100"}, "method": "POST", "signal": {}}
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [ONBOARDING] loading status loading
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [TRPC] POST deployWallet 200 in 6325ms
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] RESYNC test12, interval 10s
 LOG  [TRPC] fetching http://192.168.1.22:3000/chain/84532/getAccountHistory?batch=1&input=%7B%220%22%3A%7B%22address%22%3A%220x96b7f09c81F8510f0342aDA751094B5dc6171663%22%2C%22sinceBlockNum%22%3A5922978%7D%7D, timeout 10000ms {"body": undefined, "headers": {"content-type": "application/json", "x-daimo-platform": "ios 17.0.1", "x-daimo-version": "1.5.0 #100"}, "method": "GET", "signal": {}}
 LOG  [TRPC] GET getAccountHistory 200 in 177ms
 LOG  [SYNC] got history {"address":"0x96b7f09c81F8510f0342aDA751094B5dc6171663","name":"test12","sinceBlockNum":5922978,"lastBlock":5923521,"lastBlockTimestamp":1707615330,"lastFinalizedBlock":5922798,"lastBalance":"1050000","numTransfers":0,"numNamedAccounts":0,"numAccountKeys":1,"chainGasConstants":{"estimatedFee":0.1,"paymasterAddress":"0x0000000000dd6dd248ab5487218e1c2d7fbb29c9","maxPriorityFeePerGas":"1150000","maxFeePerGas":"1150349","preVerificationGas":"10000000"},"recommendedExchanges":[{"cta":"Bridge from any wallet","url":"https://daimo.com/bridge/test12"},{"cta":"Deposit from Coinbase","url":"https://pay.coinbase.com/buy/select-asset?appId=2be3ccd9-6ee4-4dba-aba8-d4b458fe476d&defaultExperience=send&destinationWallets=%5B%7B%22address%22%3A%220x96b7f09c81F8510f0342aDA751094B5dc6171663%22%2C%22assets%22%3A%5B%22USDC%22%5D%2C%22supportedNetworks%22%3A%5B%22base%22%5D%7D%5D"}],"suggestedActions":[{"id":"passkey-backup-new-account","title":"Secure your account","subtitle":"Keep your account safe with a passkey backup","url":"daimo://settings/add-passkey"},{"id":"2023-12-join-tg-5","icon":"comment-discussion","title":"Feedback? Ideas?","subtitle":"Join our Telegram group.","url":"https://t.me/+to2ghQJfgic0YjA9"}]}
 LOG  [SYNC] skipping sync result for 0x96b7f09c81F8510f0342aDA751094B5dc6171663. Server has finalized block 5922798 < local 5922978
 LOG  [SYNC] SUCCEEDED test12
 LOG  [ACCOUNT] save test12
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [SYNC] skipping sync, attempted sync recently
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [SYNC] RESYNC test12, interval 10s
 LOG  [TRPC] fetching http://192.168.1.22:3000/chain/84532/getAccountHistory?batch=1&input=%7B%220%22%3A%7B%22address%22%3A%220x96b7f09c81F8510f0342aDA751094B5dc6171663%22%2C%22sinceBlockNum%22%3A5922978%7D%7D, timeout 10000ms {"body": undefined, "headers": {"content-type": "application/json", "x-daimo-platform": "ios 17.0.1", "x-daimo-version": "1.5.0 #100"}, "method": "GET", "signal": {}}
 LOG  [TRPC] GET getAccountHistory 200 in 184ms
 LOG  [SYNC] got history {"address":"0x96b7f09c81F8510f0342aDA751094B5dc6171663","name":"test12","sinceBlockNum":5922978,"lastBlock":5923527,"lastBlockTimestamp":1707615342,"lastFinalizedBlock":5922978,"lastBalance":"1050000","numTransfers":0,"numNamedAccounts":0,"numAccountKeys":1,"chainGasConstants":{"estimatedFee":0.1,"paymasterAddress":"0x0000000000dd6dd248ab5487218e1c2d7fbb29c9","maxPriorityFeePerGas":"1150000","maxFeePerGas":"1150349","preVerificationGas":"10000000"},"recommendedExchanges":[{"cta":"Bridge from any wallet","url":"https://daimo.com/bridge/test12"},{"cta":"Deposit from Coinbase","url":"https://pay.coinbase.com/buy/select-asset?appId=2be3ccd9-6ee4-4dba-aba8-d4b458fe476d&defaultExperience=send&destinationWallets=%5B%7B%22address%22%3A%220x96b7f09c81F8510f0342aDA751094B5dc6171663%22%2C%22assets%22%3A%5B%22USDC%22%5D%2C%22supportedNetworks%22%3A%5B%22base%22%5D%7D%5D"}],"suggestedActions":[{"id":"passkey-backup-new-account","title":"Secure your account","subtitle":"Keep your account safe with a passkey backup","url":"daimo://settings/add-passkey"},{"id":"2023-12-join-tg-5","icon":"comment-discussion","title":"Feedback? Ideas?","subtitle":"Join our Telegram group.","url":"https://t.me/+to2ghQJfgic0YjA9"}]}
 LOG  [SYNC] synced test12 0x96b7f09c81F8510f0342aDA751094B5dc6171663.
{"oldBlock":5923517,"oldBalance":"1.05","oldTransfers":2,"newBlock":5923527,"newBalance":"1.05","newTransfers":2,"nPending":0}
 LOG  [SYNC] SUCCEEDED test12
 LOG  [ACCOUNT] save test12
 LOG  [APP] rendering
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [APP] rendering
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [ONBOARDING] chainId baseSepolia
 LOG  [ONBOARDING] chainId baseSepolia
 ```

### LINKED ISSUE #703: NFC link serves as invite
First part of #694

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Closes #703 -- https://streamable.com/yxymeq

Closes #695 for now -- we should definitely pull the logic out of the hook, weird race conditions between useExisting and createNew, tested by running through onboarding a few times

### CODE REVIEW

File: apps/daimo-mobile/src/action/actStatus.ts [hunk 1]
```diff
@@ -43,10 +43,10 @@ export function useActStatus() {
       }
       const elapsedMs = Date.now() - startTime.current;
       console.log(
-        `[ACTION] ${elapsedMs}ms: ${as.status} > ${status} ${message}`
+        `[ACTION] ${name} - ${elapsedMs}ms: ${as.status} > ${status} ${message}`
       );
       if (status !== "loading") {
-        console.log(`[ACTION] ${status}, total time ${elapsedMs}ms`);
+        console.log(`[ACTION] ${name} - ${status}, total time ${elapsedMs}ms`);
```
Comment (dcposch) [inline apps/daimo-mobile/src/action/actStatus.ts:49]: 👍

File: apps/daimo-mobile/src/action/useExistingAccount.ts [hunk 2]
```diff
@@ -29,6 +30,8 @@ export function useExistingAccount(
 
   // TODO: Does TRPC have a better way to "watch" an endpoint?
   useEffect(() => {
+    if (alreadyCreating) return; // Disable use existing check if already creating
```
Comment (dcposch) [inline apps/daimo-mobile/src/action/useExistingAccount.ts:33]: guessing this is the race condition fix?

File: apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx [hunk 3]
```diff
@@ -73,22 +68,9 @@ export function InvitePage({
...
-      const inviteStatus = linkStatus.data as DaimoInviteStatus;
-      setIsValid(inviteStatus.isValid);
-      setSender(undefined); // TODO: Add senders to invite codes
-    } else {
-      setIsValid(false);
-      setSender(undefined);
-    }
+    const inviteStatus = getInviteStatus(linkStatus.data);
+    setIsValid(inviteStatus.isValid);
+    setSender(inviteStatus.sender);
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx:73]: 🚀 nice cleanup

File: apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx [hunk 4]
```diff
@@ -77,7 +75,7 @@ export default function OnboardingScreen({
   // During onboarding, listen for payment or invite link invites
   useEffect(() => {
     getInitialURLOrTag(true).then((url) => {
-      if (url == null) return;
+      if (!url) return;
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx:78]: intentional change?

`!url` is a slightly confusing statement roughly equivalent to `url == null || url === ""` = null, undefined, or ""

File: apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx [hunk 5]
```diff
@@ -86,8 +84,6 @@ export default function OnboardingScreen({
     return () => subscription.remove();
   }, []);
 
-  console.log(`[ONBOARDING] chainId ${daimoChain}`);
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx:89]: i see, it was just printing on every render..

File: apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx [hunk 6]
```diff
@@ -116,8 +112,14 @@ export default function OnboardingScreen({
     setName("");
     setInviteLink(undefined);
     createReset && createReset();
+    setStartedCreating(false);
   };
 
+  useEffect(() => {
+    if (page === "create") setStartedCreating(true);
+    if (page === "intro") setStartedCreating(false);
+  }, [page]);
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx:121]: hmmm adding a latching state machine thing on which onboarding page you're on feels like... more moving parts.

full list
"create"
  | "existing"
  | "new-allow-notifications"
  | "existing-allow-notifications"
  | "new-loading";
  
  could do something stateless like this?
  
  ```
  const startedCreating = page !== "create" && !page.startsWith("new-")
  ```

File: packages/daimo-api/src/api/deployWallet.ts [hunk 7]
```diff
@@ -108,11 +100,9 @@ export async function deployWallet(
...
 
   const explorer = chainConfig.chainL2.blockExplorers!.default.url;
-  const inviteMeta =
-    inviteLinkStatus?.link.type +
-    (inviteLinkStatus?.link.type === "invite"
-      ? ` ${inviteLinkStatus.link.code}`
-      : "");
+  const inviteMeta = inviteLinkStatus
+    ? formatDaimoLink(inviteLinkStatus.link)
+    : "old version";
```
Comment (dcposch) [inline packages/daimo-api/src/api/deployWallet.ts:105]: 🚀

File: packages/daimo-common/src/invite.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  DaimoInviteStatus,
+  DaimoLinkStatus,
+  DaimoNoteState,
+  DaimoNoteStatus,
+  DaimoRequestStatus,
+  DaimoRequestV2Status,
+} from "./daimoLinkStatus";
+import { EAccount } from "./eAccount";
+
+/* Interpret text as a potential invite link */
```
Comment (dcposch) [inline packages/daimo-common/src/invite.ts:12]: great call putting this all in one place

**Review by dcposch — APPROVED** [review]
LGreatTM

File: apps/daimo-mobile/src/action/useExistingAccount.ts (hunk 2 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/action/useExistingAccount.ts:33]: yep

File: apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx (hunk 4 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx:78]: ah not intentional; reverted

File: apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx (hunk 6 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/onboarding/OnboardingScreen.tsx:121]: fair, done

**Pull request merged**